File: FunctionalPatterns/Structural/Decorator.fs
namespace FunctionalPatterns
namespace FunctionalPatterns.Structural

module Decorator =

    type Window = { draw: unit -> unit }

    let newNormalWindow () =
        { draw = fun () -> printfn "A normal window" }

    let verticalScrollbar window =
        { draw = window.draw >> (fun () -> printfn "  with scroll bar") }

    let menuBar window =
        { draw = window.draw >> (fun () -> printfn "  with menu bar")}

File: ObjectOrientedPatterns/Structural/Decorator.cs
namespace ObjectOrientedPatterns.Structural.Decorator
{
    // see https://en.wikipedia.org/wiki/Decorator_pattern

    public interface IWindow
    {
        public void Draw();
    }

    public class NormalWindow : IWindow
    {
        public void Draw() => Console.WriteLine("A normal window");
    }

    public class WindowDecorator : IWindow
    {
        public WindowDecorator(IWindow window)
        {
            myWindow = window;
        }
        public virtual void Draw() => myWindow.Draw();
        private IWindow myWindow;
    }

    public class VerticalScrollbarDecorator : WindowDecorator
    {
        public VerticalScrollbarDecorator(IWindow window) : base(window) { }
        public override void Draw()
        {
            base.Draw();
            Console.WriteLine("  with a vertical scroll bar");
        }
    }

    public class MenuBarDecorator : WindowDecorator
    {
        public MenuBarDecorator(IWindow window) : base(window) { }
        public override void Draw()
        {
            base.Draw();
            Console.WriteLine("  with a menu bar");
        }
    }
}
